core/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -225,6 +248,31 @@ public CompiledScript compile(Script script, ScriptContext scriptContext, Map<St
     }
 
     /**
+     * Check whether there have been too many compilations within the last minute, throwing a circuit breaking exception if so.
+     */
+    public void checkCompilationLimit() {
+        long now = System.nanoTime();
```

Comment:
Yeah, it's a variant of the [token bucket](https://en.wikipedia.org/wiki/Token_bucket) algorithm